### DISCUSSION TITLE
Sanitize field values or input field values

### DISCUSSION THREAD

**Discussion by donbeave** [graphql-java/graphql-java#3027] [Ideas]
Hello!

First of all, thank you for developing such an amazing engine. It's beautiful work and a very good innovation that will move the Java ecosystem forward.

I found one issue, but not sure if is it by design, or it's lack of my knowledge.

I have such DSL:

```graphql
directive @stripHtml on ARGUMENT_DEFINITION | INPUT_FIELD_DEFINITION

type Query {
  searchProfile(contains: String! @stripHtml, limit: Int): [Profile!]!
}

type Mutation {
  signUp(input: SignUpInput!): SignUpPayload!
}

input SignUpInput {
  username: String! @stripHtml
  password: String!
  firstName: String!
  lastName: String!
}

type Profile {
  username: String!
  fullName: String!
}
```

I would like to use `@stripHtml` directive to sanitize values, like in `searchProfile` the `contains` argument must remove all HTML code from the value, and the same for the field `username` in `SignUpInput`.

Looks like the current version of `graphql-java` allows doing so by using custom data fetchers only, and sanitizing values customly inside the data fetcher. But this isn't an easy way, as it requires going through all mapped entities and checking does it inside a list of fields or arguments annotated with this directive.

But I'm looking for an API that allows doing so somehow before, and with the sanitized results already in `DataFetchingEnvironment` class, like this method will return sanitized values already:

https://github.com/graphql-java/graphql-java/blob/477d9a08965831b8a8f14245b428a79336d04624/src/main/java/graphql/schema/DataFetchingEnvironment.java#L47.

And in terms of `SignUpInput`, let's imagine I'm using a plain Java DTO class for that input, and the sanitized value will be passed to the method `setUsername(String username)`.

@andimarek @bbakerman What do you think, is it possible to add some processor API that will allow us to post-process the value?

**Comment by bbakerman** [comment]
I understand what you want and we have considered a kind of pre-validation callback before but every time we considered it - we realised two things

* its hard to know what things needs to be validated / changed in the layer
* we already have a callback layer called the DataFetcher

So really we aim for the data fetcher to be the callback layer - graphql-java is complicated enough without need more sub specialised callbacks

There are a couple of things you can do here.

If you are building your schema from SDL (and it looks like you do) then you can use `graphql.schema.idl.SchemaDirectiveWiring` to wire in new data fetchers that do custom things first and the call the old data fetcher

Take a look at https://github.com/graphql-java/graphql-java-extended-validation which uses this.

```
directive @Size(min : Int = 0, max : Int = 2147483647, message : String = "graphql.validation.Size.message")
                        on ARGUMENT_DEFINITION | INPUT_FIELD_DEFINITION

    input Application {
        name : String @Size(min : 3, max : 100)
    }

    type Query {
        hired (applications : [Application!] @Size(max : 10)) : [Boolean]
    }
```

If you squint it a bit like yours - except it stops on validation errors - yours wants to change the argument values

It uses directives on fields as validation rules and introduces a wrapper data fetcher that enforces the rules.  In this case it only validates.  It does not modify the input args but it could

Imagine this `SchemaDirectiveWiring`

```
    @Override
    public GraphQLFieldDefinition onField(SchemaDirectiveWiringEnvironment<GraphQLFieldDefinition> env) {
        GraphQLFieldsContainer fieldsContainer = env.getFieldsContainer();
        GraphQLFieldDefinition fieldDefinition = env.getFieldDefinition();

        // you need to check the applied directives on the field args and their types to see
        // if you custom sanitizing applies - if not leave it alone
        // you will need to descend into the types of the arguments
        if (! hasYourDirectives(fieldDefinition, fieldDefinition.getArguments())) {
            return  fieldDefinition;
        }

        final DataFetcher<?> originalDF = env.getCodeRegistry().getDataFetcher(fieldsContainer, fieldDefinition);
        final DataFetcher<?> newDF = originalEnv -> {
            Map<String, Object> originalArguments = originalEnv.getArguments();
            Map<String, Object> newArguments = sanitizeArgs(originalEnv, originalArguments);

            DataFetchingEnvironment newEnv = DataFetchingEnvironmentImpl.newDataFetchingEnvironment(originalEnv).arguments(newArguments).build();
            return originalDF.get(newEnv);
        };

        env.getCodeRegistry().dataFetcher(fieldsContainer, fieldDefinition, newDF);

        return fieldDefinition;
    }
```

This  needs to do 2 things - look at the field and its arguments the type of the arguments for your custom directives - if they are present then continue

Then this creates a new DataFetcher for the field whose job it is to sanitize the arguments and create a new DFE with those args.

It then calls the original data fetcher who is none the wiser that code aspects have been wired in.

This schema wiring happens are schema build time (once) and so the originally wired data fetcher is still called - but with different values

Let me know if this helps or you have more questions.

**Comment by bbakerman** [comment]
ps this is more a discussion than an bug so I am moving it to discussions

**Comment by paullarionov** [comment]
Hello Brad @bbakerman and Alexey @donbeave,

Thank you both for **great question** and **brilliant answer**. 

I tried `SchemaDirectiveWiring` solution above and find out that `onField` is used for output types, but for input types (like `Application` in your example) there will be called other method `onInputObjectField`:

```
    @Override
    public GraphQLInputObjectField onInputObjectField(SchemaDirectiveWiringEnvironment<GraphQLInputObjectField> env) {
        GraphQLFieldsContainer fieldsContainer = env.getFieldsContainer(); // this is null
        GraphQLFieldDefinition fieldDefinition = env.getFieldDefinition(); // this is null
        GraphQLInputObjectField inputObjectField = env.getElement();

        if (!hasYourDirectives(inputObjectField)) {
            return inputObjectField;
        }

        // there is null in fieldsContainer and fieldDefinition
        // that is why we cannot override DataFetcher in CodeRegistry
        
        // final DataFetcher<?> originalDF = env.getCodeRegistry().getDataFetcher(fieldsContainer, fieldDefinition);
        // final DataFetcher<?> newDF = originalEnv -> {
        //     Map<String, Object> originalArguments = originalEnv.getArguments();
        //     Map<String, Object> newArguments = sanitizeArgs(originalEnv, originalArguments);
        //     DataFetchingEnvironment newEnv = DataFetchingEnvironmentImpl.newDataFetchingEnvironment(originalEnv).arguments(newArguments).build();
        //     return originalDF.get(newEnv);
        // };
        // env.getCodeRegistry().dataFetcher(fieldsContainer, fieldDefinition, newDF);

        return inputObjectField;
    }
```

For input objects there is no `GraphQLFieldsContainer` and `GraphQLFieldDefinition`, that is why it is not possible to override `DataFetcher` in `DataFetchingEnvironment`. I have concern that data fetchers are simply not applicable for input objects.

Guys, do you have some ideas?

**Reply by bbakerman** [reply]
`GraphQLInputObjectField` are indeed not suitable for data fetchers - only output "fields" are ever fetched"

However input fields and input types represent the  shape of the field argument data

```
query q {
    someOutputField(args : { compoundName : { first : "Brad", last : "Baker" }})
}
```

In the above the output field arguments might have come from some type system like

```
type Query {
   someOutputField( args : CompundName!) : String
}

input CompundName {
   first : String  @stripHtml
   last : String
}
```

So only `someOutputField` is ever fetched and hence has a data fetcher - however the input types and input type fields represent the compound data that can come in as arguments

One of the challenges the original question prompted was they wanted to "walk the input" type system looking for a directive (`@stripHtml` and then changing the data.

In the example given it has a String input arg (easy - 1 level deep) and the more complex arg called `SignUpInput`

```
type Mutation {
  signUp(input: SignUpInput!): SignUpPayload!
}

input SignUpInput {
  username: String! @stripHtml
  password: String!
  firstName: String!
  lastName: String!
}

```

now this is more complex because you need to walk the data (the input data coerced object) AND you need to walk the input types at the same time to work out if there is a directive on a field

eg imagine this input data in memory (its not a type system - its input data in JSON5 format say)

```

  {
    username: '<b>stripMe</b>',
    password: 'hunter2',
    firstName: 'Brad',
    lastName: 'Baker',
  },
```

you need to get the `input` argument and the coerced value for that which I have shown above, walk to each layer of the `SignUpInput` while walking down the input data object - if it has the directive - change the  input object some how

Off hand I dont know if graphql-java has a "visitor" that can walk BOTH types and data - we have a  `graphql.schema.SchemaTraverser` that can traverse the schema elements (eg output fields -> arguments -> input types -> input type fields) but I dont know of one that will also descend the input data objects say

It might be handy to have such a thing

**Reply by donbeave** [reply]
> Off hand I dont know if graphql-java has a "visitor" that can walk BOTH types and data - we have a graphql.schema.SchemaTraverser that can traverse the schema elements (eg output fields -> arguments -> input types -> input type fields) but I dont know of one that will also descend the input data objects say
> 
> It might be handy to have such a thing

Agree, I think a such visitor will be very helpful, especially for library developers, who are using `graphql-java` as an engine and providing an abstraction layer on top of it, like [graphql-java-tools](https://github.com/graphql-java-kickstart/graphql-java-tools).

**Reply by bbakerman** [reply]
I had a go at a input value traverser

https://github.com/graphql-java/graphql-java/pull/3033

Its modelled on this use case

**Reply by bbakerman** [reply]
https://github.com/graphql-java/graphql-java/pull/3033/files#diff-5a348698d5fd54e955c08a6f7b4364c74eeb08a4ef8310e089fa3138b99825e0R458

has an integration test like what I mentioned above

**Reply by donbeave** [reply]
That's amazing! Thank you so much, @bbakerman! This is exactly what I'm looking for.

**Reply by AVLOUIS99** [reply]
How can I use this, is there an applicable example?

**Reply by bbakerman** [reply]
The link above leads to a PR and that PR is merged - its in the graphql code base since Dec 3, 2022

The PR has a test and the test above talks about it in a general sense.

**Reply by AVLOUIS99** [reply]
Thanks for the quick reply and thanks for the extension which I had been looking for a while!

I saw that and tried to familiarize myself with it, but my problem goes back to the question of [paullarionov](https://github.com/graphql-java/graphql-java/discussions/3027#discussioncomment-4238854), because unfortunately I have exactly the same problem and had hoped the thread could help me with it. Thanks in advance!

I am relatively new to graphql and would appreciate some information.

**Reply by paullarionov** [reply]
Hi @AVLOUIS99. Sure, here is small code example of a directive, which traverses input objects:

```
import graphql.schema.DataFetcher;
import graphql.schema.DataFetchingEnvironment;
import graphql.schema.GraphQLFieldDefinition;
import graphql.schema.GraphQLFieldsContainer;
import graphql.schema.idl.SchemaDirectiveWiring;
import graphql.schema.idl.SchemaDirectiveWiringEnvironment;

@Component
public class YourDirectiveWiring implements SchemaDirectiveWiring {

    @Autowired YourValueTraverser yourValueTraverser;

    @Override
    public GraphQLFieldDefinition onField(SchemaDirectiveWiringEnvironment<GraphQLFieldDefinition> env) {
        GraphQLFieldsContainer fieldsContainer = env.getFieldsContainer();
        GraphQLFieldDefinition fieldDefinition = env.getFieldDefinition();

        final DataFetcher<?> originalDF = env.getCodeRegistry().getDataFetcher(fieldsContainer, fieldDefinition);
        final DataFetcher<?> newDF = originalEnv -> {
            DataFetchingEnvironment newEnv = ValueTraverser.visitPreOrder(originalEnv, yourValueTraverser);
            return originalDF.get(newEnv);
        };

        env.getCodeRegistry().dataFetcher(fieldsContainer, fieldDefinition, newDF);

        return fieldDefinition;
    }

}
```

Where `YourValueVisitor` is the implementation of `ValueTraverser` (https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/analysis/values/ValueTraverser.java)

**Reply by AVLOUIS99** [reply]
Hi @paullarionov thanks for the reply!
Isn't the onField method only for field_definition directives? 
```graphql
input typeFilter {
    state: Int
    opening: String @uppercase
}

directive @uppercase on INPUT_FIELD_DEFINITION
```
This is the last problem of understanding that I still have in this matter.

**Reply by AVLOUIS99** [reply]
Got it! Thank you guys! 👍

**Reply by bjoernmayer** [reply]
Hey @AVLOUIS99
I am stuck with the same problem. How did you solve it?
At a first glance, I see no usecase for the ValueTraverser

**Reply by alirezatali** [reply]
i have the same problem but can you put a YourValueTraverser  sample?
